libs/blockscout-auth/src/lib.rs

```diff
@@ -0,0 +1,140 @@
+use cookie::Cookie;
+use serde::Serialize;
+use std::collections::HashMap;
+use thiserror::Error;
+use tonic::{codegen::http::header::COOKIE, metadata::MetadataMap, Request};
+
+const JWT_TOKEN_NAME: &str = "_explorer_key";
+const CSRF_TOKEN_NAME: &str = "_csrf_token";
+
+#[derive(Debug)]
+pub struct AuthSuccess {
+    pub user_id: String,
+}
+
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum Error {
+    #[error("invalid data: {0}")]
+    InvalidData(String),
+    #[error("user is unauthorized: {0}")]
+    Unauthorized(String),
+    #[error("blockscout invalid response: {0}")]
+    BlockscoutApiError(String),
+}
+
+pub async fn auth_from_tonic<T: Serialize>(
+    request: Request<T>,
+    blockscout_host: &str,
+) -> Result<AuthSuccess, Error> {
+    let jwt = extract_jwt(request.metadata())?;
+    let csrf_token = serde_json::to_value(request.into_inner())
+        .map_err(|e| Error::InvalidData(format!("invalid request payload: {e}")))?
+        .get(CSRF_TOKEN_NAME)
+        .and_then(|token| token.as_str().map(|s| s.to_string()));
+    auth_from_tokens(jwt.as_ref(), csrf_token.as_deref(), blockscout_host).await
+}
+
+pub async fn auth_from_tokens(
+    jwt: &str,
+    csrf_token: Option<&str>,
+    _blockscout_host: &str,
+) -> Result<AuthSuccess, Error> {
+    // TODO: replace with actual blockscout api call
+    let mut jwt = jwt.to_string();
+    if let Some(cstf_token) = csrf_token {
+        jwt.push_str(cstf_token)
+    }
+    Ok(AuthSuccess { user_id: jwt })
+}
+
+fn extract_jwt(metadata: &MetadataMap) -> Result<String, Error> {
+    let cookies = get_cookies(metadata)?;
+    let token = cookies
+        .get(JWT_TOKEN_NAME)
+        .map(|cookie| cookie.value())
+        .ok_or_else(|| Error::InvalidData(format!("'{JWT_TOKEN_NAME}' not found in request")))?;
+    Ok(token.to_string())
+}
+
+fn get_cookies(metadata: &MetadataMap) -> Result<HashMap<String, Cookie>, Error> {
+    let cookies_raw = metadata
+        .get(COOKIE.as_str())
+        .map(|value| value.to_str())
+        .ok_or_else(|| Error::InvalidData("no cookies were provided".to_string()))?
+        .map_err(|e| Error::InvalidData(format!("invalid cookie format: {e}")))?;
```

Comment:
Nit: may be a little bit clearer:

```suggestion
    let cookies_raw = metadata
        .get(COOKIE.as_str())
        .ok_or_else(|| Error::InvalidData("no cookies were provided".to_string()))?
        .to_str()
        .map_err(|e| Error::InvalidData(format!("invalid cookie format: {e}")))?;
```